Panic when generating interface for redis client

I love the idea of this project! Unfortunately this is what I hit when I tried it out:

`> interfacer -for \"github.com/go-redis/redis\".Client -as mock.redis`

```
panic: internal: t=*types.Signature, orig=*types.Signature

goroutine 1 [running]:
panic(0x217c40, 0x429cf5550)
	$GOROOT/src/runtime/panic.go:500 +0x1a1
github.com/rjeczalik/interfaces.(*Type).setFromType(0x430661878, 0x367400, 0x425b33290, 0x0, 0x367400, 0x425b33290)
	$GOPATH/src/github.com/rjeczalik/interfaces/type.go:65 +0x350
github.com/rjeczalik/interfaces.newType(0x42596d1d0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	$GOPATH/src/github.com/rjeczalik/interfaces/type.go:29 +0x8b
github.com/rjeczalik/interfaces.buildInterfaceForPkg(0x4204b20b0, 0x430661eb0, 0x7fff5fbfe4e1, 0x19, 0x42582d818, 0x1, 0x0)
	$GOPATH/src/github.com/rjeczalik/interfaces/interface.go:140 +0x41c
github.com/rjeczalik/interfaces.buildInterface(0x430661eb0, 0x420433dd8, 0x95de4, 0x7fff5fbfe4e1, 0x21, 0x266bba)
	$GOPATH/src/github.com/rjeczalik/interfaces/interface.go:84 +0x2c0
github.com/rjeczalik/interfaces.NewWithOptions(0x430661eb0, 0x21, 0x4204a8060, 0x0, 0x0, 0x4204a2000)
	$GOPATH/src/github.com/rjeczalik/interfaces/interface.go:45 +0x156
main.main()
	$GOPATH/src/github.com/rjeczalik/interfaces/cmd/interfacer/main.go:73 +0xf0
```

Am I using it wrong or is there a bug here?

@jenspinney What go version are you using and what's the HEAD of your local redis repo?

I'm getting:

```
$ go version
go version go1.8.3 darwin/amd64
```

```
src/github.com/go-redis/redis $ git rev-parse HEAD
eb066030c08c254136a0dbfe9f7a3625a0871beb
```

```
$ interfacer  -for \"github.com/go-redis/redis\".Client -as mock.redis
panic: recursive types not supported: github.com/go-redis/redis.Client

goroutine 1 [running]:
github.com/rjeczalik/interfaces.collectMethods(0xc426d599d8, 0xc4243cb170, 0x81, 0x136c320, 0xc4236c0c00)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/interface.go:160 +0x331
github.com/rjeczalik/interfaces.collectMethods(0xc426d599d8, 0xc4243cb350, 0x80, 0x136c320, 0xc4236c0c00)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/interface.go:176 +0x22e
github.com/rjeczalik/interfaces.collectMethods(0xc426d599d8, 0xc422dab740, 0x7f, 0x136c320, 0xc4236c0c00)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/interface.go:176 +0x22e
...
```

Ok, with #5 I'm able to reproduce the problem:

```
$ interfacer  -for \"github.com/go-redis/redis\".Client -as mock.Redis
panic: internal: t=*types.Signature, orig=*types.Signature

goroutine 1 [running]:
github.com/rjeczalik/interfaces.(*Type).setFromType(0xc421f1b8f8, 0x136c3a0, 0xc4246e0420, 0x0, 0x136c3a0, 0xc4246e0420)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/type.go:65 +0x2f1
github.com/rjeczalik/interfaces.newType(0xc422110960, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/type.go:29 +0x89
github.com/rjeczalik/interfaces.buildInterfaceForPkg(0xc4200de0b0, 0xc421f1bf20, 0x7fff5fbfe23b, 0x19, 0xc423fd2e58, 0x127a901, 0x0)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/interface.go:140 +0x41d
github.com/rjeczalik/interfaces.buildInterface(0xc421f1bf20, 0xc4200cbe18, 0x11f510c, 0x7fff5fbfe23b, 0x21, 0x126d6c9)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/interface.go:84 +0x227
github.com/rjeczalik/interfaces.NewWithOptions(0xc4200cbf20, 0x21, 0xc42000a800, 0x0, 0x0, 0xc4200aa200)
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/interface.go:45 +0x151
main.main()
        /Users/rjeczalik/src/github.com/rjeczalik/interfaces/cmd/interfacer/main.go:73 +0x105
```

👀

--- a/interface.go
+++ b/interface.go
@@ -138,9 +138,11 @@ func buildInterfaceForPkg(pkg *loader.PackageInfo, opts *Options) (Interface, er
 		}
 		for i := range fn.Ins {
 			fn.Ins[i] = newType(ins.At(i))
+			fixup(&fn.Ins[i], opts.Query)
 		}
 		for i := range fn.Outs {
 			fn.Outs[i] = newType(outs.At(i))
+			fixup(&fn.Outs[i], opts.Query)
 		}
 		inter = append(inter, fn)
 	}

--- a/type.go
+++ b/type.go
@@ -51,6 +51,8 @@ func (typ *Type) setFromType(t types.Type, depth int, orig types.Type) {
 		typ.setFromStruct(t)
 	case *types.Named:
 		typ.setFromNamed(t)
+	case *types.Signature:
+		typ.setFromSignature(t)
 	case *types.Pointer:
 		if depth == 0 {
 			typ.IsPointer = true
@@ -79,6 +81,12 @@ func (typ *Type) setFromInterface(t *types.Interface) {
 }
 
 func (typ *Type) setFromStruct(t *types.Struct) {
+	if typ.Name == "" {
+		typ.Name = t.String()
+	}
+}
+
+func (typ *Type) setFromSignature(t *types.Signature) {
 	if typ.Name == "" {
 		typ.Name = t.String()
 	}

--- a/type.go
+++ b/type.go
@@ -3,6 +3,8 @@ package interfaces
 import (
 	"fmt"
 	"go/types"
+	"path"
+	"strings"
 )
 
 // Type is a simple representation of a single parameter type.
@@ -12,6 +14,7 @@ type Type struct {
 	ImportPath  string `json:"importPath,omitempty"`  // import path of the package
 	IsPointer   bool   `json:"isPointer,omitempty"`   // whether the parameter is a pointer
 	IsComposite bool   `json:"isComposite,omitempty"` // whether the type is map, slice, chan or array
+	IsFunc      bool   `json:"isFunc,omitempty"`      // whether the type if function
 }
 
 // String gives Go code representation of the type.
@@ -52,6 +55,7 @@ func (typ *Type) setFromType(t types.Type, depth int, orig types.Type) {
 	case *types.Named:
 		typ.setFromNamed(t)
 	case *types.Signature:
+		typ.IsFunc = true
 		typ.setFromSignature(t)
 	case *types.Pointer:
 		if depth == 0 {
@@ -111,4 +115,17 @@ func (typ *Type) setFromComposite(t compositeType, depth int, orig types.Type) {
 		typ.Name = t.String()
 	}
 	typ.setFromType(t.Elem(), depth+1, orig)
+}
+
+func fixup(typ *Type, q *Query) {
+	// Hacky fixup for renaming:
+	//
+	//   GeoAdd(string, []*github.com/go-redis/redis.GeoLocation) *redis.IntCmd
+	//
+	// to:
+	//
+	//   GeoAdd(string, []*redis.GeoLocation) *redis.IntCmd
+	//
+	// Should be fixed layer below, in type.go.
+	typ.Name = strings.Replace(typ.Name, q.Package, path.Base(q.Package), -1)
 }

